security/oc_sdi.c

```diff
@@ -55,20 +55,45 @@ oc_sec_sdi_init(void)
 #endif /* OC_DYNAMIC_ALLOCATION */
 }
 
+static void
+sec_sdi_free(oc_sec_sdi_t *sdi)
+{
+  oc_free_string(&sdi->name);
+}
+
+#ifdef OC_HAS_FEATURE_DEVICE_ADD
+
 void
-oc_sec_sdi_free(void)
+oc_sec_sdi_init_at_index(size_t device_index, bool needs_realloc)
 {
-#ifdef OC_DYNAMIC_ALLOCATION
-  if (g_sdi == NULL) {
-    return;
+  if (needs_realloc) {
+    size_t device_count = oc_core_get_num_devices();
+    assert(device_index == device_count - 1);
+    oc_sec_sdi_t *sdi =
+      (oc_sec_sdi_t *)realloc(g_sdi, device_count * sizeof(oc_sec_sdi_t));
+    if (sdi == NULL) {
+      oc_abort("Insufficient memory");
+    }
+    g_sdi = sdi;
+  } else {
+    sec_sdi_free(&g_sdi[device_index]);
   }
-#endif /* OC_DYNAMIC_ALLOCATION */
+  memset(&g_sdi[device_index], 0, sizeof(oc_sec_sdi_t));
```

Comment:
I saw same pattern at lot's places -> is it possible to move it to generic function ?
``
void
oc_instantiate_object_at_index(void** array, item_size size_t, index size_t, bool needs_realloc)
```